controllers/nnfcontainerprofile_helpers.go

```diff
@@ -0,0 +1,169 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	dwsv1alpha1 "github.com/HewlettPackard/dws/api/v1alpha1"
+	nnfv1alpha1 "github.com/NearNodeFlash/nnf-sos/api/v1alpha1"
+)
+
+// findProfileToUse verifies a NnfContainerProfile named in the directive or verifies that a default can be found.
+func findContainerProfileToUse(ctx context.Context, clnt client.Client, args map[string]string) (*nnfv1alpha1.NnfContainerProfile, error) {
+	var profileName string
+
+	NnfContainerProfile := &nnfv1alpha1.NnfContainerProfile{}
+
+	profileNamespace := os.Getenv("NNF_CONTAINER_PROFILE_NAMESPACE")
+
+	// If a profile is named then verify that it exists.  Otherwise, verify
+	// that a default profile can be found.
+	profileName, present := args["profile"]
+	if present == false {
+		NnfContainerProfiles := &nnfv1alpha1.NnfContainerProfileList{}
+		if err := clnt.List(ctx, NnfContainerProfiles, &client.ListOptions{Namespace: profileNamespace}); err != nil {
+			return nil, err
+		}
+		profilesFound := make([]string, 0, len(NnfContainerProfiles.Items))
+		for _, profile := range NnfContainerProfiles.Items {
+			if profile.Data.Default {
+				objkey := client.ObjectKeyFromObject(&profile)
+				profilesFound = append(profilesFound, objkey.Name)
+			}
+		}
+		// Require that there be one and only one default.
+		if len(profilesFound) == 0 {
+			return nil, fmt.Errorf("Unable to find a default NnfContainerProfile to use")
+		} else if len(profilesFound) > 1 {
+			return nil, fmt.Errorf("More than one default NnfContainerProfile found; unable to pick one: %v", profilesFound)
+		}
+		profileName = profilesFound[0]
+	}
+	if len(profileName) == 0 {
+		return nil, fmt.Errorf("Unable to find an NnfContainerProfile name")
+	}
+	err := clnt.Get(ctx, types.NamespacedName{Namespace: profileNamespace, Name: profileName}, NnfContainerProfile)
+	if err != nil {
+		return nil, err
+	}
+	return NnfContainerProfile, nil
+}
+
+// findPinnedProfile finds the specified pinned profile.
+func findPinnedContainerProfile(ctx context.Context, clnt client.Client, namespace string, pinnedName string) (*nnfv1alpha1.NnfContainerProfile, error) {
+
+	NnfContainerProfile := &nnfv1alpha1.NnfContainerProfile{}
+	err := clnt.Get(ctx, types.NamespacedName{Namespace: namespace, Name: pinnedName}, NnfContainerProfile)
+	if err != nil {
+		return nil, err
+	}
+	if !NnfContainerProfile.Data.Pinned {
+		return nil, fmt.Errorf("Expected pinned NnfContainerProfile, but it was not pinned: %s", pinnedName)
+	}
+	return NnfContainerProfile, nil
+}
+
+// createPinnedProfile finds the specified profile and makes a pinned copy of it.
+func createPinnedContainerProfile(ctx context.Context, clnt client.Client, clntScheme *runtime.Scheme, args map[string]string, owner metav1.Object, pinnedName string) (*nnfv1alpha1.NnfContainerProfile, error) {
```

Comment:
What, no generics?

Reply:
oh.   Go for it, Nate.

Reply:
Eventually, yes :)

I did put a TODO in the top of the file. I don't like having to duplicate all this, either.

Reply:
```go
type NnfContainerProfile struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`
}
​
func (in *NnfContainerProfile) DeepCopyInto(out *NnfContainerProfile) {}
func (in *NnfContainerProfile) DeepCopy() *NnfContainerProfile        { return nil }
func (in *NnfContainerProfile) DeepCopyObject() runtime.Object        { return nil }
​
type Profile interface {
	client.Object
	nnfv1alpha1.NnfStorageProfile | NnfContainerProfile
}
​
func findProfileToUseGeneric[P Profile](ctx context.Context, c client.Client) (*P, error) {
​
	var profile P
​
	if err := c.Get(ctx, types.NamespacedName{Name: "test", Namespace: "default"}, profile); err != nil {
		return nil, err
	}
​
	return &profile, nil
}
```

Reply:
Those deepcopy's should already be in zz_generated.deepcopy.go.